decisions/20240313-try-w3up.md

```diff
@@ -0,0 +1,142 @@
+
+## Risks
+
+### degrading nft.storage's existing functionality
+
+We wouldn't want these changes to have any negative impact on users that rely on nft.storage's current setup. We only want to enable evaluation of new ways of delivering on that functionality that are seamless for end-users.
+
+#### Mitigation: feature switch for initial rollout
+
+We'll deploy this functionality to use web3.storage initially behind an opt-in feature switch. Once deployed, most uploads will not use the web3.storage functionality. Only accounts we add to a configured allow-list will get persisted to w3up. This removes risk of everyday users being impacted by bugs in the new code, while still enabling employees we allowlist to test the code in production. After testing in production, we can add more to the allowlist (e.g. perhaps nft.storage employees), or we can remove the feature switch to roll out the functionality for all users of nft.storage.
+
+### Are there any backwards-incompatible changes?
+
+No
+
+### Does this project have special implications for security and data privacy?
+
+Some minor implications.
+After these changes, uploads will be shared with web3.storage in a new way. However, my understanding is that nft.storage is for public data, and this data is already shared with web3.storage through other means for the purpose of storage/serving on ipfs/filecoin. So there are no new data sharing concerns.
+
+### Could this change significantly increase load on any of our backend systems?
+
+Unlikely. As I understand it, the status quo for file uploads on nft.storage is that it is sent the full file in an http request, and then must store it. There may be a more work to do in some cases to store it (encoding/signing/invoking web3.storage), but this shouldn't be a **significant** increase in load. In the worst case, it may be 100% more cpu used for each file uploads, but we can mitigate the rollout of that via feature switch, and based on testing we can find ways of optimizing it.
+
+### Does this project have any dependencies?
+
+#### dependency on collaboration with nft.storage
+
+This project can't succeed without a sponsor from nft.storage and willingness to try out the proposed changes to the nft.storage codebase in a production-like environment. If we have this, we can help prove out nft.storage storing data via web3.storage and relying on web3.storage productized services for serving on ipfs and filecoin dealmaking.
+
+## Alternative Solutions
+
+I don't know of any alternate solutions I can link to that provide a path to servicing nft.storage's web3 storage needs without relying on dagcargo. If you know of some, please add them here.
```

Comment:
use w3filecoin directly. But that seems already out of the table :)

Reply:
@vasco-santos Do you think using w3filecoin directly might be a more promising approach than what we describe here? It can be on the table, I just don't understand what that would entail and pros/cons.

Reply:
Here is the proposal for it I wrote a way back and discussed with Alan and Hannah https://hackmd.io/@vasco-santos/Skxt0yd5a/edit

w3filecoin was designed for multiple storefronts, so w3up/others, and my idea was to not go through w3up to avoid mixing the data at that level. And this together with the separate aggregation process for both. The general sense though is that using w3up makes more sense on a perspective on having nft.storage as a user of w3up in the long run. But if the expectation we have is it to be 3 months, making it a client of w3filecoin, while continuing to write to old bucket and so on, would make more sense to me.

But after talking with Alan, Hannah, Reid, we opted on the w3up lane, and that is why I say that it seems out of the table. I generally agree with using w3up because it is simpler if we won't care about splitting aggregation + have nft.storage in the long run as a client of w3up

Reply:
I added a section for this alternate solution. Thanks for sharing it and context. 🙏  https://github.com/nftstorage/nft.storage/pull/2518/commits/b3c22cdb8e70be29f4b6cf66f70e6f9f4c24e07e